[FSDP] Add HSDP parity unit test

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106080
* #106068
* #106034
* __->__ #106131
* #106072
* #106033
* #106032

With >= 4 GPUs:
```
python -m pytest test/distributed/fsdp/test_fsdp_hybrid_shard.py -k test_fsdp_hybrid_shard_parity
```

Differential Revision: [D47852458](https://our.internmc.facebook.com/intern/diff/D47852458)

Commit:
[FSDP] Add HSDP parity unit test

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add HSDP parity unit test"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add HSDP parity unit test"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Add HSDP parity unit test"




[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47852458).

rohan-varma commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -313,11 +317,73 @@ def patched_collective(orig_collective, counter, *args, **kwargs):
+        fsdp_optim = torch.optim.Adam(fsdp_model.parameters(), lr=1e-2)
+        hsdp_optim = torch.optim.Adam(hsdp_model.parameters(), lr=1e-2)
+        for _ in range(5):
+            inp = fsdp_model.module.get_input(torch.device("cuda"))
```
you might have to seed differently across ranks to ensure that inputs are diff across ranks for this test to be nontrivial?

awgu replied:
Fixed!

Commit:
Update on "[FSDP] Add HSDP parity unit test"


Differential Revision: [D47852458](https://our.internmc.facebook.com/intern/diff/D47852458)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add HSDP parity unit test"


Differential Revision: [D47852458](https://our.internmc.facebook.com/intern/diff/D47852458)

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
